[Core] Prevent exception propagation

feltech opened the pull request:
Part of #15. USD cannot handle exceptions being thrown during stage resolution and will terminate the application (sigabrt/segfault).

Best guess so far for the expected error behaviour is to return a default-constructed object (e.g. empty string, invalid timestamp, etc).

So add and use a decorator that will catch all exceptions and log them before returning a default-constructed object.

feltech committed:
[Core] Prevent propagation of exceptions

Part of #15. USD cannot handle exceptions being thrown during stage
resolution and will terminate the application (sigabrt/segfault).

Best guess so far for the expected error behaviour is to return a
default-constructed object (e.g. empty string, invalid timestamp, etc).

So add and use a decorator that will catch all exceptions and log them
before returning a default-constructed object.

Signed-off-by: David Feltell <david.feltell@foundry.com>

foundrytom commented on tests/test_resolver.py:
```diff
@@ -101,6 +101,20 @@ def test_non_assetized_child_ref_assetized_grandchild():
+    logs = capfd.readouterr()
+    assert (
+        "OpenAssetIO error in UsdOpenAssetIOResolver::_GetExtension: RuntimeError: 130:"
+        " Malformed BAL reference: Missing entity name in path component" in logs.err
```
[consider] Is there a middle ground here where we're not coupling to BALs specific messages? Just we've had this  a few times before where we create a lot of maintenance churn, say when BAL changes its error messages.

feltech replied:
How about leaving the `RuntimeError: 130`, but removing the actual message?  It's still somewhat tied to BAL, but then, the whole test relies on knowledge of BAL's quirks to trigger a specific failure mode.

feltech replied:
Removed the message part in b6eb77d0

foundrytom replied:
Nice one, like it 👍

foundrytom reviewed (approved):
Awesome sauce 👍

elliotcmorris commented on src/resolver.cpp:
```diff
@@ -114,8 +114,7 @@ Ref locationInManagerContextForEntity(const openassetio::hostApi::ManagerPtr &ma
           // TODO(DF): Better conversion of BatchElementError to
           //  appropriate exception type.
-          std::string errorMsg = "OpenAssetIO error code ";
-          errorMsg += std::to_string(static_cast<int>(error.code));
```
[Question] Why the removal of this, it seems like decent context?

feltech replied:
The new exception wrapper already adds an "OpenAssetIO error" prefix, so it seemed redundant.  I guess we can bring back the "error code" part if you think it's worthwhile?

feltech replied:
Reintroduced "error code" prefix in 270784c8

elliotcmorris replied:
Ah, i didn't see you'd get an error : prefix. I agree that it was fine before, but it's equally fine now as well, so whatever's least work. Sorry for the runaround.

elliotcmorris commented on src/resolver.cpp:
```diff
@@ -133,14 +132,58 @@ Ref locationInManagerContextForEntity(const openassetio::hostApi::ManagerPtr &ma
+ * @return Result of callable if no exception occurred, otherwise a
+ * default-constructed object.
+ */
+template <typename Fn>
+auto catchAndLogExceptions(Fn &&fn, const openassetio::log::LoggerInterfacePtr &logger,
+                           const std::string_view name) {
```
:+1: Awesome solution

elliotcmorris commented on src/resolver.cpp:
```diff
@@ -133,14 +132,58 @@ Ref locationInManagerContextForEntity(const openassetio::hostApi::ManagerPtr &ma
+    msg += ": unknown non-exception type caught";
+    logger->critical(msg);
+  }
+  return std::invoke_result_t<Fn>{};
```
[Question] Is simply calling `fn();` off the table? I suspect this is perhaps something to do with handling the `void` case?

feltech replied:
This is default-constructing an instance of the return type of `fn`, without actually calling `fn` (since that throws an exception if we've got here).

elliotcmorris replied:
Ah, duh!

elliotcmorris commented on src/resolver.cpp:
```diff
@@ -168,54 +211,69 @@ UsdOpenAssetIOResolver::~UsdOpenAssetIOResolver() {
-  logger_->debug("  result: " + identifier);
-
-  return identifier;
+  const std::string fnName = TF_FUNC_NAME();
```
[Consider] Rather than capturing a name variable, we could just log the function name on entry and avoid the extra state. If we use the `catchAndLogExceptions` return value, I think we can avoid mixing the logic and the logging altogether.

A good formulation might be something like

```cpp
logger_->debug("OPENASSETIO_RESOLVER: " + TF_FUNC_NAME())
logger_->debug("  assetPath: " + assetPath);
logger_->debug("  anchorAssetPath: " + anchorAssetPath.GetPathString());

auto result = catchAndLogExceptions(...){...}

logger_->debug("  result: " + identifier);
```

That might give you scope to do some early returns in some if/else blocks in the implementation. It's quite nice actually, creates a seperation between system-logic and behaviour loggin.

feltech replied:
I like it. My only concern is multithreading, where log lines can be interleaved, so you don't know for certain which method caused the exception.

foundrytom replied:
[fix] Btw, David - I think you highlighted that Log message should be multi-line anyway, not split into separate calls? Now we're using `logger` not `TF` how about just dumping it into one entry?

feltech replied:
How does 81adf29 look?

foundrytom replied:
mar vel ouse

elliotcmorris replied:
Looks great.

elliotcmorris reviewed (approved):
Nice one. Just a couple of questions.

feltech committed:
[CI] Show pytest's captured log output in console

Part of #15. Now that we have stdout/err -capturing tests, it's handy
to see the output in the CI job log.

Signed-off-by: David Feltell <david.feltell@foundry.com>

feltech committed:
[Core] Improve debug logging

Part of #15. Replace existing debug logs with two log calls per
member function, one on function entry and one on exit.

Logging entry into a function allows us to detect any function that
causes an error/hang.

Logging exit from a function allows us to log the return value.

Log the thread ID so that interleaved logs can be detected.

Remove "OPENASSETIO_RESOLVER" prefix, since that is duplicated in
the function name, which includes the class name.

Signed-off-by: David Feltell <david.feltell@foundry.com>

foundrytom reviewed (approved):
FFTSQMAYL
